### ISSUE TITLE
[basic.lifetime] The concept of "storage is released" is underspecified

### ISSUE THREAD

**Issue by xmh0511** [cplusplus/CWG#99]
Full name of submitter (unless configured in github; will be published with the issue): Jim X

[basic.life] p7 says
> Similarly, before the lifetime of an object has started but after the storage which the object will occupy has been allocated or, after the lifetime of an object has ended and before the storage which the object occupied is reused or **released**
>> - the glvalue is used to access the object, or
>> - the glvalue is used to call a non-static member function of the object, or
>> - the glvalue is bound to a reference to a virtual base class ([dcl.init.ref]), or
>> - the glvalue is used as the operand of a dynamic_­cast ([expr.dynamic.cast]) or as the operand of typeid.

In what situation is the storage considered as having been released? It is not clear. Consider this example
````cpp
struct A{
   A(){} 
  ~A(){}
  A& rf = *this;
};
A& getA(){
   A a;
  return a;
}
A& r = getA();
r.rf;  // #1 
````
Is `#1` well-defined? The glvalue `r` refers to an object that has been destroyed. However, it does not violates any bullet in the list of [basic.life] p7. **Access** to the object is defined in [defns.access], anyway, a reference or the referenced type is not a scalar type. 

[basic.stc.general] p4 says
> When the end of the duration of a region of storage is reached, the values of all pointers representing the address of any part of that region of storage become invalid pointer values. Indirection through an invalid pointer value and passing an invalid pointer value to a deallocation function have undefined behavior. 

This rule is also not suitable for this case since there is no pointer value used here. Maybe "release" is the right constraint for this case, however, its meaning is not clear in this document. Further
````cpp
A a;
a.~A();
a.rf; // well-defined
````
Is `a.rf` well-defined an intent of the standard?

**Comment by jensmaurer** [comment]
Obviously, the storage duration of `a` has ended (it's an automatic variable); see [basic.stc.auto] p1 "The storage for these entities lasts until the block in which they are created exits."

The remaining question is whether the offset calculation for `a.rf` can be done, which is the subject of CWG1530 and CWG1853.
